servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/IdleTimeoutConnectionFilter.java

```diff
@@ -0,0 +1,242 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(IdleTimeoutConnectionFilter.class);
+
+    private static final Cancellable CANCELLED = () -> { };
+
+    private final long timeoutNs;
+    @Nullable
+    private final Executor timeoutExecutor;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param timeout timeout duration after which an idle connection is closed
+     */
+    IdleTimeoutConnectionFilter(final Duration timeout) {
+        this.timeoutNs = ensurePositive(timeout).toNanos();
+        this.timeoutExecutor = null;
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param timeout timeout duration after which an idle connection is closed
+     * @param timeoutExecutor the {@link Executor} to use for scheduling the timer notifications
+     */
+    IdleTimeoutConnectionFilter(final Duration timeout, final Executor timeoutExecutor) {
+        this.timeoutNs = ensurePositive(timeout).toNanos();
+        this.timeoutExecutor = requireNonNull(timeoutExecutor);
+    }
+
+    private static Duration ensurePositive(final Duration timeout) {
+        if (ZERO.compareTo(timeout) >= 0) {
+            throw new IllegalArgumentException("timeout: " + timeout.toNanos() + " ns (expected: >0)");
+        }
+        return timeout;
+    }
+
+    private static Executor contextExecutor(ExecutionContext<HttpExecutionStrategy> context) {
+        return context.executionStrategy().hasOffloads() ? context.executor() : context.ioExecutor();
+    }
+
+    @Override
+    public StreamingHttpConnectionFilter create(final FilterableStreamingHttpConnection connection) {
+        return new ConnectionIdleTimeoutFilterImpl(connection, timeoutNs,
+                timeoutExecutor != null ? timeoutExecutor : contextExecutor(connection.executionContext()));
+    }
+
+    @Override
+    public HttpExecutionStrategy requiredOffloads() {
+        return offloadNone();
+    }
+
+    private static final class ConnectionIdleTimeoutFilterImpl extends StreamingHttpConnectionFilter
+            implements Runnable {
+
+        private static final AtomicIntegerFieldUpdater<ConnectionIdleTimeoutFilterImpl> requestsUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(ConnectionIdleTimeoutFilterImpl.class, "requests");
+        private static final AtomicReferenceFieldUpdater<ConnectionIdleTimeoutFilterImpl, Cancellable>
+                timeoutTaskUpdater = AtomicReferenceFieldUpdater.newUpdater(ConnectionIdleTimeoutFilterImpl.class,
+                Cancellable.class, "timeoutTask");
+
+        private volatile int requests;
+        @Nullable
+        private volatile Cancellable timeoutTask;
+
+        private final long timeoutNs;
+        private final Executor timeoutExecutor;
+
+        // While it may look like "volatile" is not required for this variable for "happens-before" visibility because
+        // we always write to "requests" after updating "lastResponseTime" and read "requests" before reading
```

Comment:
`we always write to "requests" after updating "lastResponseTime"` -> this isn't quite true yet pending the comment below

Reply:
Consider simplifying/clarifying this comment to say the volatile is not for general visibility but to prevent [word tearing](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.6)